### PR TITLE
retransmits shreds recovered from erasure codes

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
Shreds recovered from erasure codes have not been received from turbine
and have not been retransmitted to other nodes downstream. This results
in more repairs across the cluster which is slower.

#### Summary of Changes
* 4th commit reworks retransmit stage to simply receive shreds instead of packets; because converting recovered "shreds" to "packets" just to send them to retransmit-stage is superfluous. This will also avoid [partially deserializing packet in retransmit stage](https://github.com/solana-labs/solana/blob/32501866b/core/src/retransmit_stage.rs#L233) when the entire packet was already deserialized to shred earlier in the window-stage.
* In preparation of the above change, first 2 commits move a number of metrics related to packet-counts from retransmit stage to window-service earlier in the pipeline. Metrics for repair and discard packets are emitted in the window stage and, with the 4th commit, those packets are no longer sent down to retransmit-stage (which was pointless since they will not be retransmitted).
* Last commit, channels through recovered shreds to retransmit stage in order to further broadcast the shreds to downstream nodes in the tree.

### CODE REVIEW

File: core/src/window_service.rs [hunk 1]
```diff
@@ -316,29 +354,32 @@ where
...
-            let _ = retransmit.send(packets);
-        }
+    // Exclude repair packets from retransmit.
+    let _ = retransmit_sender.send(
+        shreds
+            .iter()
+            .zip(&repair_infos)
+            .filter(|(_, repair_info)| repair_info.is_none())
+            .map(|(shred, _)| shred)
+            .cloned()
```
Comment (carllin) [inline core/src/window_service.rs:368]: Wonder if we could avoid this clone if we switched over to some `Arc<Shred>`, but that's a story for another time

File: core/src/retransmit_stage.rs [hunk 2]
```diff
@@ -560,21 +454,23 @@ impl RetransmitStage {
...
         completed_data_sets_sender: CompletedDataSetsSender,
-        max_slots: &Arc<MaxSlots>,
+        max_slots: Arc<MaxSlots>,
         rpc_subscriptions: Option<Arc<RpcSubscriptions>>,
         duplicate_slots_sender: Sender<Slot>,
         ancestor_hashes_replay_update_receiver: AncestorHashesReplayUpdateReceiver,
     ) -> Self {
         let (retransmit_sender, retransmit_receiver) = channel();
+        // https://github.com/rust-lang/rust/issues/39364#issuecomment-634545136
+        let _retransmit_sender = retransmit_sender.clone();
```
Comment (carllin) [inline core/src/retransmit_stage.rs:464]: ick, I've run into this too.

**Comment by carllin** [comment]
> Shreds recovered from erasure codes have not been received from turbine
and have not been retransmitted to other nodes downstream. This results
in more repairs across the cluster which is slower.

Hmm, what if in cases where we just happen to receive some combination of erasure shreds + data shreds first such that we recover some data shreds that we *eventually* would have gotten from turbine. Will this cause a massive spike in bandwidth if we then retransmit these recovered shreds, even though turbine had no problem circulating them? 

To avoid this, should we wait for a bit to see that these recovered shreds don't ultimately arrive through turbine?

**Review by carllin — APPROVED** [review]
Nice, this is so much cleaner, awesome! 😃

File: core/src/window_service.rs [hunk 3]
```diff
@@ -546,16 +589,16 @@ impl WindowService {
         insert_sender: CrossbeamSender<(Vec<Shred>, Vec<Option<RepairMeta>>)>,
         verified_receiver: CrossbeamReceiver<Vec<Packets>>,
         shred_filter: F,
-        leader_schedule_cache: Arc<LeaderScheduleCache>,
         bank_forks: Arc<RwLock<BankForks>>,
-        retransmit: PacketSender,
+        retransmit_sender: Sender<Vec<Shred>>,
```
Comment (jbiseda) [inline core/src/window_service.rs:593]: thanks for renaming these :)

**Review by jbiseda — APPROVED** [review]
looks good!

**Comment by KeKo6988** [comment]
> #### Problem
> Shreds recovered from erasure codes have not been received from turbine
> and have not been retransmitted to other nodes downstream. This results
> in more repairs across the cluster which is slower.
> 
> #### Summary of Changes
> * 4th commit reworks retransmit stage to simply receive shreds instead of packets; because converting recovered "shreds" to "packets" just to send them to retransmit-stage is superfluous. This will also avoid [partially deserializing packet in retransmit stage](https://github.com/solana-labs/solana/blob/32501866b/core/src/retransmit_stage.rs#L233) when the entire packet was already deserialized to shred earlier in the window-stage.
> * In preparation of the above change, first 2 commits move a number of metrics related to packet-counts from retransmit stage to window-service earlier in the pipeline. Metrics for repair and discard packets are emitted in the window stage and, with the 4th commit, those packets are no longer sent down to retransmit-stage (which was pointless since they will not be retransmitted).
> * Last commit, channels through recovered shreds to retransmit stage in order to further broadcast the shreds to downstream nodes in the tree.

File: core/benches/retransmit_stage.rs [hunk 4]
```diff
@@ -17,11 +17,10 @@ use {
         shred::Shredder,
     },
     solana_measure::measure::Measure,
-    solana_perf::packet::{Packet, Packets},
     solana_runtime::{bank::Bank, bank_forks::BankForks},
     solana_sdk::{
         hash::Hash,
-        pubkey,
+        pubkey::Pubkey,
```
Comment (KeKo6988) [inline core/benches/retransmit_stage.rs:23]: 5rZrRNA9NyAig5B6F8V98EfUFPxS3T6hKa4rT5cdvbJC

File: core/src/retransmit_stage.rs [hunk 5]
```diff
@@ -33,17 +33,13 @@ use {
         pubkey::Pubkey,
```
Comment (KeKo6988) [inline core/src/retransmit_stage.rs:33]: 5rZrRNA9NyAig5B6F8V98EfUFPxS3T6hKa4rT5cdvbJC

**Review by KeKo6988 — APPROVED** [review]

**Comment by behzadnouri** [comment]
> Hmm, what if in cases where we just happen to receive some combination of erasure shreds + data shreds first such that we recover some data shreds that we _eventually_ would have gotten from turbine. Will this cause a massive spike in bandwidth if we then retransmit these recovered shreds, even though turbine had no problem circulating them?
> 
> To avoid this, should we wait for a bit to see that these recovered shreds don't ultimately arrive through turbine?

That should be picked up by duplicates check in retransmit stage (i.e. `check_if_already_received`) and so the 2nd copy of the shred will be skipped:
https://github.com/solana-labs/solana/blob/0e5ea36cc/core/src/retransmit_stage.rs#L227-L253

**Pull request merged**